_posts/2021-11-29-when-estimating-a-software-project-you-should-not.md

```diff
@@ -0,0 +1,34 @@
+---
+title: When estimating a software project you should not...
+---
+
+I'm currently working on an internal GitHub project where, for a change, my estimations have been wildly off. Before this project, I'd had a good multi-year streak of never overshooting my estimates and consistently hitting deadlines. This break down, combined with some recent reading, has made me think about the myths we tell ourselves about software estimation so I figure I'd repeat some here. These are in a rough order from "most obvious" to "least obvious".
+
+When estimating a software project you should not:
+
+- refuse to estimate because of any of the reasons below
+- plan things to go as you expect
+- ignore testing, code review, deployment and integration takes time, often longer than writing the code
+- present exact dates or times without ranges
+- be pressured into different estimates for the same scope based on "business needs"
+- be unwilling to pad estimates based on them "feeling" too long
+- estimate the entire project without splitting it up individually ranged, estimated sub-projects
+- have a non-engineer produce the estimate
+- have an engineer who will not be doing the work produce the estimate, without discussion with the engineer that will be doing the work
+- assume that engineers will do the work in the same amount of time, even if they are the same "level". [Individuals matter](https://danluu.com/people-matter/).
+- assume that other individuals/teams/departments will always be as responsive as you need them to be and never block progress
+- estimate assuming that you will not reduce technical debt, leaving things better than you found them
+- ignore [Hofstadter's law](https://en.wikipedia.org/wiki/Hofstadter%27s_law): "It always takes longer than you expect, even when you take into account Hofstadter's Law."
+- ignore the sporadic relevance of [Parkinson's law](https://en.wikipedia.org/wiki/Parkinson%27s_law): "work expands so as to fill the time available for its completion"
+- ignore about [the project management triangle](https://en.wikipedia.org/wiki/Project_management_triangle) and discuss in advance which corner to optimise for
+- ignore that context-switching has a cost, particularly between different projects
+- ignore that re-orgs, team changes and internal or external controversies will distract engineers and cause projects to take longer
```

Comment:
I think many of these points could maybe be summarised to what to do: i.e. to include a contingency because anything can really happen. I don't think I could get away with adding extra time to estimates because 'there might be a re-org'....

Reply:
@DW-DW-DW Good point. Have added a follow-up nothing what you should do.